### PR TITLE
PR: Move debug cell / file / config to Debugger plugin

### PR DESCRIPTION
**Pull request by impact27**
## Description of Changes

* [ ] Wrote at least one-line docstrings (for any new functions)
* [ ] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Fixes #

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:

### CODE REVIEW

File: spyder/plugins/ipythonconsole/plugin.py [hunk 1]
```diff
@@ -634,6 +634,11 @@ def run_script(self, filename, wdir, args='', debug=False,
             True if the console namespace should be used, False otherwise.
         focus_to_editor: bool, optional
             Leave focus in the editor after execution.
+        method : str or None
+            method to run the file. must accept the same arguments as runfile.
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/plugin.py:638]:
```suggestion
            Method to run the file. It must accept the same arguments as
            `runfile`.
```

File: spyder/plugins/ipythonconsole/plugin.py [hunk 2]
```diff
@@ -634,6 +634,11 @@ def run_script(self, filename, wdir, args='', debug=False,
...
+        force_wdir: bool
+            the working directory is ignored on remote kernel except if
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/plugin.py:640]:
```suggestion
            The working directory is ignored on remote kernels except if
```

File: spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py [hunk 3]
```diff
@@ -1031,8 +1031,10 @@ def test_execute_events_dbg(ipyconsole, qtbot):
         shell.execute('%debug print()')
 
     # Set processing events to True
-    ipyconsole.set_conf('pdb_execute_events', True)
-    shell.set_pdb_execute_events(True)
+    ipyconsole.set_conf('pdb_execute_events', True, section='debugger')
+    shell.set_pdb_configuration({
+        'pdb_execute_events': True
+        })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py:1037]:
```suggestion
    })
```

File: spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py [hunk 4]
```diff
@@ -1043,8 +1045,10 @@ def test_execute_events_dbg(ipyconsole, qtbot):
     assert shell._control.toHtml().count('img src') == 1
 
     # Set processing events to False
-    ipyconsole.set_conf('pdb_execute_events', False)
-    shell.set_pdb_execute_events(False)
+    ipyconsole.set_conf('pdb_execute_events', False, section='debugger')
+    shell.set_pdb_configuration({
+        'pdb_execute_events': False
+        })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/tests/test_ipythonconsole.py:1051]:
```suggestion
    })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 5]
```diff
@@ -777,37 +777,41 @@ def change_clients_autocall(self, value):
...
     def change_clients_pdb_ignore_lib(self, value):
-        for idx, client in enumerate(self.clients):
-            self._change_client_conf(
-                client,
-                client.shellwidget.set_pdb_ignore_lib,
-                value)
+        for client in self.clients:
+            client.shellwidget.set_pdb_configuration({
+                'pdb_ignore_lib': value
+                })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:785]:
```suggestion
            })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 6]
```diff
@@ -777,37 +777,41 @@ def change_clients_autocall(self, value):
...
     def change_clients_pdb_execute_events(self, value):
-        for idx, client in enumerate(self.clients):
-            self._change_client_conf(
-                client,
-                client.shellwidget.set_pdb_execute_events,
-                value)
+        for client in self.clients:
+            client.shellwidget.set_pdb_configuration({
+                'pdb_execute_events': value
+                })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:792]:
```suggestion
            })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 7]
```diff
@@ -777,37 +777,41 @@ def change_clients_autocall(self, value):
...
     def change_clients_pdb_use_exclamation_mark(self, value):
-        for idx, client in enumerate(self.clients):
-            self._change_client_conf(
-                client,
-                client.shellwidget.set_pdb_use_exclamation_mark,
-                value)
+        for client in self.clients:
+            client.shellwidget.set_pdb_configuration({
+                'pdb_use_exclamation_mark': value
+                })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:799]:
```suggestion
            })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 8]
```diff
@@ -777,37 +777,41 @@ def change_clients_autocall(self, value):
...
     def change_clients_pdb_stop_first_line(self, value):
-        for idx, client in enumerate(self.clients):
-            self._change_client_conf(
-                client,
-                client.shellwidget.set_pdb_stop_first_line,
-                value)
+        for client in self.clients:
+            client.shellwidget.set_pdb_configuration({
+                'pdb_stop_first_line': value
+                })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:806]:
```suggestion
            })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 9]
```diff
@@ -777,37 +777,41 @@ def change_clients_autocall(self, value):
...
+
+    def set_spyder_breakpoints(self):
+        """Set Spyder breakpoints into all clients"""
+        for cl in self.clients:
+            cl.shellwidget.set_pdb_configuration({
+                'breakpoints': self.get_conf(
+                    'breakpoints', default={}, section='run')
+                })
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:814]:
```suggestion
            })
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 10]
```diff
@@ -2220,20 +2210,35 @@ def norm(text):
             client = self.get_current_client()
 
         if client is not None:
-            # Internal kernels, use runcell
-            if client.get_kernel() is not None and not run_cell_copy:
+            is_spyder_kernel = client.shellwidget.is_spyder_kernel
+            # Only use copy for runcell
+
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:2215]:
```suggestion

            # Only use copy for runcell
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 11]
```diff
@@ -2220,20 +2210,35 @@ def norm(text):
...
+                # Can not use custom function on non-spyder kernels
+                QMessageBox.warning(
+                    self,
+                    _('Warning'),
+                    _("The client is not a spyder-kernel "
+                      "to run <b>{}</b>.<br><br>Please Use a spyder-kernel."
+                      ).format(method),
+                    QMessageBox.Ok
+                )
+                return
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:2241]: Why this change? If a kernel is not a Spyder one, we can still run code on it by simply pasting it, which is what I think we are doing here, aren't we?

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 12]
```diff
@@ -2287,12 +2286,17 @@ def run_script(self, filename, wdir, args, debug, post_mortem,
...
+                method = "runfile"
             # If spyder-kernels, use runfile
             if client.shellwidget.is_spyder_kernel:
-                line = "%s('%s'" % ('debugfile' if debug else 'runfile',
-                                    norm(filename))
+                line = method + "('%s'" % (norm(filename))
                 if args:
                     line += ", args='%s'" % norm(args)
+                if (wdir and client.shellwidget.is_external_kernel
+                        and not force_wdir):
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:2297]:
```suggestion
                if (
                    wdir and client.shellwidget.is_external_kernel
                    and not force_wdir
                ):
```

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 13]
```diff
@@ -2301,16 +2305,27 @@ def run_script(self, filename, wdir, args, debug, post_mortem,
...
+            else:
+                QMessageBox.warning(
+                    self,
+                    _('Warning'),
+                    _("The client is not a spyder-kernel "
+                      "to run <b>{}</b>.<br><br>Please Use a spyder-kernel."
+                      ).format(method),
+                    QMessageBox.Ok
+                )
+                return
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:2328]: I don't understand why you added this. Precisely the previous `elif` takes care of non-Spyder kernels, as the comment below it clearly says.

**Review by ccordoba12 — COMMENTED** [review]
Thanks a lot @impact27 for your help with this!

File: spyder/plugins/ipythonconsole/widgets/main_widget.py (hunk 11 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:2241]: Yes but not debug, so if it is not a spyder and not tu file, we can not run

File: spyder/plugins/ipythonconsole/widgets/main_widget.py (hunk 13 above)
Comment (impact27) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:2330]: The reason is that I plan to add a profile_file function (and maybe others) so if it is a spyder-kernels, no problem, if runfile, we can run on ipython, if debug file, we can run with -d. For profile file, nothing works

File: spyder/plugins/debugger/plugin.py [hunk 14]
```diff
@@ -15,14 +18,18 @@
 from spyder.plugins.debugger.widgets.main_widget import (
     DebuggerWidget)
 from spyder.api.shellconnect.mixins import ShellConnectMixin
+from spyder.utils.qthelpers import MENU_SEPARATOR
+from spyder.config.manager import CONF
+from spyder.plugins.mainmenu.api import ApplicationMenus
+from spyder.plugins.debugger.widgets.main_widget import DebuggerWidgetActions
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:24]:
```suggestion
from spyder.plugins.debugger.widgets.main_widget import DebuggerToolbarActions
```

I think the `Debug file` and `Debug cell` actions should be part of a different enum class.

File: spyder/plugins/debugger/plugin.py [hunk 15]
```diff
@@ -58,12 +65,41 @@ def on_preferences_teardown(self):
...
+        widget = self.get_widget()
+
+        widget.edit_goto.connect(editor.load)
+        widget.sig_debug_file.connect(self.debug_file)
+        widget.sig_debug_cell.connect(self.debug_cell)
+
+        names = [
+            DebuggerWidgetActions.DebugCurrentFile,
+            DebuggerWidgetActions.DebugCurrentCell,
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:77]:
```suggestion
        names = [
            DebuggerToolbarActions.DebugCurrentFile,
            DebuggerToolbarActions.DebugCurrentCell,
        ]
```

File: spyder/plugins/debugger/plugin.py [hunk 16]
```diff
@@ -58,12 +65,41 @@ def on_preferences_teardown(self):
...
+        widget = self.get_widget()
+
+        widget.edit_goto.disconnect(editor.load)
+        widget.sig_debug_file.disconnect(self.debug_file)
+        widget.sig_debug_cell.disconnect(self.debug_cell)
+
+        names = [
+            DebuggerWidgetActions.DebugCurrentFile,
+            DebuggerWidgetActions.DebugCurrentCell,
+            ]
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:99]:
```suggestion
        names = [
            DebuggerToolbarActions.DebugCurrentFile,
            DebuggerToolbarActions.DebugCurrentCell,
        ]
```

File: spyder/plugins/debugger/plugin.py [hunk 17]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
         self.get_widget().sig_show_namespace.disconnect(
             self.show_namespace_in_variable_explorer)
 
+    @on_plugin_available(plugin=Plugins.MainMenu)
+    def on_main_menu_available(self):
+        widget = self.get_widget()
+        debug_file_action = widget.get_action(
+            DebuggerWidgetActions.DebugCurrentFile)
+        debug_cell_action = widget.get_action(
+            DebuggerWidgetActions.DebugCurrentCell)
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:120]:
```suggestion
        debug_file_action = widget.get_action(
            DebuggerToolbarActions.DebugCurrentFile)
        debug_cell_action = widget.get_action(
            DebuggerToolbarActions.DebugCurrentCell)
```

File: spyder/plugins/debugger/plugin.py [hunk 18]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
...
+        mainmenu = self.get_plugin(Plugins.MainMenu)
+
+        mainmenu.remove_item_from_application_menu(
+            DebuggerWidgetActions.DebugCurrentFile,
+            menu_id=ApplicationMenus.Debug
+        )
+        mainmenu.remove_item_from_application_menu(
+            DebuggerWidgetActions.DebugCurrentCell,
+            menu_id=ApplicationMenus.Debug
+        )
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:139]:
```suggestion
        mainmenu.remove_item_from_application_menu(
            DebuggerToolbarActions.DebugCurrentFile,
            menu_id=ApplicationMenus.Debug
        )
        mainmenu.remove_item_from_application_menu(
            DebuggerToolbarActions.DebugCurrentCell,
            menu_id=ApplicationMenus.Debug
        )
```

File: spyder/plugins/debugger/plugin.py [hunk 19]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
...
     # ------------------------------------------------------------------------
+    @Slot()
+    def debug_file(self):
+        """
+        Debug current script.
+        Should only be called when an editor is avilable.
+        """
+        editor = self.get_plugin(Plugins.Editor)
+        editor.switch_to_plugin()
+        editor.run_file(method="debugfile")
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:151]:
```suggestion
        editor = self.get_plugin(Plugins.Editor, error=False)
        if editor:
            editor.switch_to_plugin()
            editor.run_file(method="debugfile")
```

File: spyder/plugins/debugger/plugin.py [hunk 20]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
...
+            menu_id=ApplicationMenus.Debug
+        )
+
     # ---- Public API
     # ------------------------------------------------------------------------
+    @Slot()
+    def debug_file(self):
+        """
+        Debug current script.
+        Should only be called when an editor is avilable.
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:147]:
```suggestion
        Debug current file.

        It should only be called when an editor is available.
```

File: spyder/plugins/debugger/plugin.py [hunk 21]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
...
+        editor = self.get_plugin(Plugins.Editor)
+        editor.switch_to_plugin()
+        editor.run_file(method="debugfile")
+
+    @Slot()
+    def debug_cell(self):
+        '''
+        Debug Current cell.
+        Should only be called when an editor is avilable.
+        '''
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:158]:
```suggestion
        """
        Debug current cell.

        It should only be called when an editor is available.
        """
```

File: spyder/plugins/debugger/plugin.py [hunk 22]
```diff
@@ -75,8 +111,54 @@ def on_variable_explorer_teardown(self):
...
+        editor = self.get_plugin(Plugins.Editor)
+        editor.run_cell(method="debugcell")
```
Comment (ccordoba12) [inline spyder/plugins/debugger/plugin.py:160]:
```suggestion
        editor = self.get_plugin(Plugins.Editor, error=False)
        if editor:
            editor.run_cell(method="debugcell")
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 23]
```diff
@@ -42,6 +42,9 @@ class DebuggerWidgetActions:
     ToggleCaptureLocals = 'toggle_capture_locals_action'
     ToggleLocalsOnClick = 'toggle_show_locals_on_click_action'
 
+    DebugCurrentFile = 'debug file'
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:45]:
```suggestion

class DebuggerToolbarActions:
    DebugCurrentFile = 'debug file'
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 24]
```diff
@@ -76,6 +79,10 @@ class DebuggerWidget(ShellConnectMainWidget):
     # Signals
     edit_goto = Signal((str, int, str), (str, int, str, bool))
     sig_show_namespace = Signal(dict, object)
+    sig_debug_file = Signal()
+    """This signal is emitted to request the current file to be debugged."""
+    sig_debug_cell = Signal()
+    """This signal is emitted to request the current cell to be debugged."""
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:85]:
```suggestion

    sig_debug_file = Signal()
    """This signal is emitted to request the current file to be debugged."""

    sig_debug_cell = Signal()
    """This signal is emitted to request the current cell to be debugged."""
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 25]
```diff
@@ -185,6 +192,24 @@ def setup(self):
             register_shortcut=True
         )
 
+        self.create_action(
+            DebuggerWidgetActions.DebugCurrentFile,
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:196]:
```suggestion
            DebuggerToolbarActions.DebugCurrentFile,
```

File: spyder/plugins/debugger/widgets/main_widget.py [hunk 26]
```diff
@@ -185,6 +192,24 @@ def setup(self):
...
+            text=_("&Debug file"),
+            tip=_("Debug file"),
+            icon=self.create_icon('debug'),
+            triggered=self.sig_debug_file,
+            register_shortcut=True,
+        )
+
+        self.create_action(
+            DebuggerWidgetActions.DebugCurrentCell,
```
Comment (ccordoba12) [inline spyder/plugins/debugger/widgets/main_widget.py:205]:
```suggestion
            DebuggerToolbarActions.DebugCurrentCell,
```

**Review by ccordoba12 — COMMENTED** [review]
Second round of reviews for you @impact27.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py [hunk 27]
```diff
@@ -2220,20 +2210,35 @@ def norm(text):
...
-                line = code.strip()
+                # Can not use custom function on non-spyder kernels
+                QMessageBox.warning(
+                    self,
+                    _('Warning'),
+                    _("The client is not a spyder-kernel "
+                      "to run <b>{}</b>.<br><br>Please Use a spyder-kernel."
+                      ).format(method),
+                    QMessageBox.Ok
+                )
```
Comment (ccordoba12) [inline spyder/plugins/ipythonconsole/widgets/main_widget.py:2240]:
```suggestion
                client.shellwidget.append_html_message(
                    _("The console is not running a Spyder-kernel, so it "
                      "can't execute <b>{}</b>.<br><br>"
                      "Please use a Spyder-kernel for this.").format(method),
                    msg_type='error',
                    before_prompt=True
                )
```

I think it's better if we use for this the API to show messages in our consoles instead of a `QMessageBox`.

File: spyder/plugins/ipythonconsole/widgets/main_widget.py (hunk 13 above)
Comment (ccordoba12) [reply spyder/plugins/ipythonconsole/widgets/main_widget.py:2330]:
```suggestion
            else:
                client.shellwidget.append_html_message(
                    _("The console is not running a Spyder-kernel, so it "
                      "can't execute <b>{}</b>.<br><br>"
                      "Please use a Spyder-kernel for this.").format(method),
                    msg_type='error',
                    before_prompt=True
                )
                return
```

Ok, thanks for the clarification @impact27. Then I suggest using the API for messages in `shellwidget` for this.

**Review by ccordoba12 — COMMENTED** [review]
Last review @impact27 then this should be ready.

**Review by ccordoba12 — APPROVED** [review]
Thanks @impact27!
